### DISCUSSION TITLE
Better scalable setup for Cypress e2e and Component Tests for large projects

### DISCUSSION THREAD

**Discussion by mirobo** [nrwl/nx#19999] [Feature Requests]
TLDR
We have a heavily customized NX setup in order to address the issue that Cypress tests in "app-e2e" do not scale as the project grows larger. We would like to see a solution that embraces splitting libraries into smaller chunks and allows to add e2e tests on a per-library basis (ideally without the need for a lot of boilerplate configuration files).

We started a project about 3 years ago with NX, Angular and Cypress. We were eager to write quite a lot of e2e tests and jest tests for components (and later Cypress component tests). While the test base grew it became obvious that NX doesn't scale when it comes to test execution because the "app-e2e" library just executes ALL the tests and doesn't care which part of the application was tested.

We started to create more Cypress CTs to counter this issue and to get the build times down and we moved (as good as possible) e2e tests into the specific libraries (we now have about 160 libraries which are included in the main app and the main app doesn't have much code other than some basic routing and lazy loading of the other modules).

I'd like to see a standard setup where a library has a Cypress Component Testing setup if Cypress is installed in package.json and also provide a e2e testing setup for each library. We handcrafted this and we can profit quite effectively from NX affected this way.

But we have a complex setup with TONS of configuration files just to have the Cypress setup (and everything else) working for each library.

The best solution would be to add some more NX affected magic to it so the e2e-tests could live in it's own directory with a single setup and NX would crawl through the test files, selectors, page-objects and determine which tests need to be run based on the code that changed in a particular library. Basically I think it could also be an option to give the e2e tests some "hints" to which components are used, by basically importing Angular components and within the page objects maybe access the components selector (not sure how other companies solve this issue but normally HOW the selector is named is totally irrelevant; it's just a means for the purpose (being able to verify that a specific component was loaded).
